Minor: Zerops Deployment Guide (#7973)
withastro/docs: src/content/docs/en/guides/deploy/zerops.mdx

type: deploy
i18nReady: true
---
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'
import { Steps } from '@astrojs/starlight/components';

[Zerops](https://zerops.io/) is a dev-first cloud platform that can be used to deploy an SSR Astro site.

This guide will walk you through deploying an Astro project using the Node.js adapter to Zerops.

## Prerequisites

- **An Astro project using the [`@astrojs/node` SSR adapter](/en/guides/integrations-guide/node/)** 
- **A Zerops account** - If you don’t already have one, you can [create a Zerops account](https://zerops.io/) for free.

:::tip[Start from a template]
The [Zerops x Astro - Node.js example app](https://github.com/zeropsio/recipe-astro-nodejs) can be imported directly into your [Zerops Dashboard](https://app.zerops.io/dashboard/projects), and deployed in one click!

```yaml
project:
  name: astro
services:
  - hostname: astronode
    type: nodejs@20
    buildFromGit: https://github.com/zeropsio/recipe-astro-nodejs
    ports:
      - port: 4321
        httpSupport: true
    enableSubdomainAccess: true
    minContainers: 1
```
:::

## Creating a Zerops Node.js project

> You can create a Node.js service for your Astro site through the [Zerops `project add` wizard](https://app.zerops.io/dashboard/project-add) wizard, or by importing an Astro site using `.yaml`.

Reviewer: ```suggestion
You can create a Node.js service for your Astro site through the [Zerops `project add` wizard](https://app.zerops.io/dashboard/project-add), or by importing an Astro site using `.yaml`.
```
wizard wizard!

---
title: Deploy your Astro Site to Zerops
description: How to deploy your Astro site to the web using Zerops.
type: deploy
i18nReady: true
---

> import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'

Reviewer: ```suggestion
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'
import { Steps } from '@astrojs/starlight/components';
```

This guide will walk you through deploying an Astro Node.js service to Zerops.

## Prerequisites

- A Zerops account. If you don’t already have one, you can create an account for free.
- `@astrojs/node` package

:::tip[How about a one-click example?]
Import [Zerops x Astro - Node.js](https://github.com/zeropsio/recipe-astro-nodejs) example app into Zerops.
Open [Zerops Dashboard](https://app.zerops.io/dashboard/projects), click "Import project", paste ⬇️ — built, deployed, done.
```yaml
project:
  name: astro
services:
  - hostname: astronode
    type: nodejs@20
    buildFromGit: https://github.com/zeropsio/recipe-astro-nodejs
    ports:
      - port: 4321
        httpSupport: true
    enableSubdomainAccess: true
    minContainers: 1
```
:::

### Creating your own Zerops project with Astro apps

You can either add project and your first Node.js service through the [project add](https://app.zerops.io/dashboard/project-add) wizard in Zerops app, or quickly by importing it using a yaml structure.

```yaml
project:
  name: my-astro-sites
services:
  - hostname: hellothere
    type: nodejs@20
    ports:
      - port: 4321
        httpSupport: true
    minContainers: 1
```

> This will setup a project called `my-astro-sites`, with a Node.js v20 service called `hellothere`, which will be ready for you to deploy your Astro app into. One Zerops project can contain many Astro apps.

Reviewer: ```suggestion
```
We put descriptions of code *above*, so that people know what they are going to read, and don't have to scroll back up to see what you meant!

> ### Setting Up Node.js SSR in your app
> Zerops is using [`@astrojs/node`](/en/guides/integrations-guide/node/) package to run the app. Add it to your deps, if you haven't already, and modify your Astro config to utilize it.
> 
> Here are the necessary lines in your `astro.config.mjs` file:
> ```js title="astro.config.mjs"
>   import { defineConfig } from 'astro/config';
>   import node from "@astrojs/node";
> 
>   export default defineConfig({
>     output: 'server',
>     adapter: node({
>       mode: "standalone"
>     }),
>   });
> ```
> 

Reviewer: ```suggestion
```
This was in prerequisites, so we don't need it here! We can assume it's already done.

## Building and deploying your app to Zerops

> Now that you you've set up your app and prepared a Node.js service on Zerops, we can trigger the build and deploy pipeline on Zerops.

Reviewer: ```suggestion
Now that you've prepared a Node.js service on Zerops,  you will need to create a `zerops.yml` file at the root of your project to trigger the build and deploy pipeline on Zerops.
```

> Start by describing how to build and run your app through `zerops.yml` file placed at the root of your codebase. `setup:` part should match the hostname your chose for your service, in case of our example - `hellothere`.

Reviewer: ```suggestion
The following example shows configuring the required build and run operations for the example project with hostname `hellothere`:
```

  <PackageManagerTabs>
    <Fragment slot="npm">
    ```yaml
      zerops:
        - setup: hellothere
          build:
            base: nodejs@20
            buildCommands:

>               - npm i
>               - npm run build
>             deploy:
>               - dist
>               - package.json
>               - node_modules
>             cache:
>               - node_modules
>               - package-lock.json
>           run:
>             start: node dist/server/entry.mjs
>           envVariables:
>             HOST: 0.0.0.0
>             NODE_ENV: production
>     ```
>     </Fragment>
>     <Fragment slot="pnpm">
>     ```yaml
>       zerops:
>         - setup: hellothere
>           build:
>             base: nodejs@20
>             buildCommands:
>               - pnpm i
>               - pnpm run build
>             deploy:
>               - dist
>               - package.json
>               - node_modules
>             cache:
>               - node_modules
>               - pnpm-lock.json
>           run:
>             start: node dist/server/entry.mjs
>           envVariables:
>             HOST: 0.0.0.0
>             NODE_ENV: production
>     ```
>     </Fragment>
>     <Fragment slot="yarn">
>     ```yaml
>       zerops:
>         - setup: astronode
>           build:
>             base: nodejs@20
>             buildCommands:
>               - yarn
>               - yarn build
>             deploy:
>               - dist
>               - package.json
>               - node_modules
>             cache:
>               - node_modules
>               - yarn.lock
>           run:
>             start: node dist/server/entry.mjs
>           envVariables:
>             HOST: 0.0.0.0
>             NODE_ENV: production

Reviewer: ``````suggestion
    <Fragment slot="npm">
    ```yaml title="zerops.yml"
      zerops:
        - setup: hellothere
          build:
            base: nodejs@20
            buildCommands:
              - npm i
              - npm run build
            deploy:
              - dist
              - package.json
              - node_modules
            cache:
              - node_modules
              - package-lock.json
          run:
            start: node dist/server/entry.mjs
          envVariables:
            HOST: 0.0.0.0
            NODE_ENV: production
    ```
    </Fragment>
    <Fragment slot="pnpm">
    ```yaml title="zerops.yml"
      zerops:
        - setup: hellothere
          build:
            base: nodejs@20
            buildCommands:
              - pnpm i
              - pnpm run build
            deploy:
              - dist
              - package.json
              - node_modules
            cache:
              - node_modules
              - pnpm-lock.json
          run:
            start: node dist/server/entry.mjs
          envVariables:
            HOST: 0.0.0.0
            NODE_ENV: production
    ```
    </Fragment>
    <Fragment slot="yarn">
    ```yaml title="zerops.yml"
      zerops:
        - setup: astronode
          build:
            base: nodejs@20
            buildCommands:
              - yarn
              - yarn build
            deploy:
              - dist
              - package.json
              - node_modules
            cache:
              - node_modules
              - yarn.lock
          run:
            start: node dist/server/entry.mjs
          envVariables:
            HOST: 0.0.0.0
            NODE_ENV: production
``````
I think we can make each of these also display the file title at the top, which is just another nice reminder about where this code goes. We'll check the deploy preview and make sure this looks right!

    ```
    </Fragment>
  </PackageManagerTabs>

> ### Trigger the pipeline by connecting the service to GitHub / GitLab
> To setup continuous deployment on either push to branch or a new release, go to your Node.js service detail and connect your Zerops service with either GitHub or GitLab repository.

Reviewer: ```suggestion
### Trigger the pipeline using GitHub / GitLab
To setup continuous deployment on either a push to a branch or on a new release, go to your Node.js service detail and connect your Zerops service with a GitHub or GitLab repository.
```

### Trigger the pipeline Using Zerops CLI (zcli)

> 1. Install Zerops CLI
> ```shell
> # To download the zcli binary directly,
> # use https://github.com/zeropsio/zcli/releases
> npm i -g @zerops/zcli
> ```
> 
> 2. Open [`Settigs > Access Token Management`](https://app.zerops.io/settings/token-management) in Zerops app, generate a new access token
> 
> 3. Login
> ```shell
> zcli login <token>
> ```
> 
> 4. Navigate to your apps root (where zerops.yml is also placed)
> ```shell
> zcli push
> ```

Reviewer: ```suggestion
<Steps>
1. Install the Zerops CLI.
		```shell
		# To download the zcli binary directly,
		# use https://github.com/zeropsio/zcli/releases
		npm i -g @zerops/zcli
		```

2. Open [`Settings > Access Token Management`](https://app.zerops.io/settings/token-management) in the Zerops app and generate a new access token.

3. Log in using your access token with the following command:
		```shell
		zcli login <token>
		```

4. Navigate to the root of your app (where `zerops.yml` is located) and run the following command to trigger the deploy:
		```shell
		zcli push
		```
</Steps>
```
Fixing indentation so code blocks align with the numbers, and using our `<Steps>` component!

## Resources

> - [Deploying Astro to Zerops in 3 mins](https://medium.com/@arjun.js/how-to-deploy-astro-to-zerops-4230816a62b4)
> - [Detailed guide to create a Node.js service](https://docs.zerops.io/nodejs/how-to/create)
> - [Managing Environment Variables](https://docs.zerops.io/nodejs/how-to/env-variables)
> - [Updating Internal Ports](https://docs.zerops.io/nodejs/how-to/ports)

Reviewer: ```suggestion
- [Deploying Astro to Zerops in 3 mins](https://medium.com/@arjun.js/how-to-deploy-astro-to-zerops-4230816a62b4)
- [Detailed guide to creating a Zerops Node.js service](https://docs.zerops.io/nodejs/how-to/create)
```
We don't need to link to every little thing in the Zerops docs! the blog post and the creation guide are great, and then people will use Zerops docs to find whatever else they might need.
